File: all-submissions/0200-number-of-islands/solution.java
class Solution {

    class State {
        int row;
        int col;
        State(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }
    int[][] directions = new int[][] {{0,1}, {1,0}, {0,-1}, {-1,0}};
    

    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        boolean[][] seen = new boolean[m][n];
        int answer = 0;
        Queue<State> queue = new LinkedList<>();
        for(int r=0; r<m;r++) {
            for (int c=0; c<n; c++) {
                if(grid[r][c] == '1' && !seen[r][c]) {
                    answer++;
                    queue.add(new State(r,c));
                    seen[r][c] = true;
                    while(!queue.isEmpty()) {
                        State state = queue.poll();
                        int row = state.row;
                        int col = state.col;
                        for(int i=0 ;i<directions.length; i++) {
                            int nextRow = row+directions[i][0];
                            int nextCol = col+directions[i][1];
                            if(isValid(nextRow, nextCol, grid) && !seen[nextRow][nextCol]) {
                                seen[nextRow][nextCol] = true;
                                queue.add(new State(nextRow, nextCol));
                            }
                        }
                    }
                }
            }
        }
        return answer;
    }

    public boolean isValid(int row, int col, char[][] grid) {
        return (row >=0 && row < grid.length && col >=0 && col < grid[0].length && grid[row][col] == '1');
    }
}

File: all-submissions/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # visited = set()
        # ROWS = len(grid)
        # COLS = len(grid[0])

        # def dfs(r, c):
        #     if r < 0 or r >=ROWS or c <0 or c >= COLS:
        #         return
        #     if grid[r][c] == "1" and (r,c) not in visited:
        #         visited.add((r,c))
        #         dfs(r+1,c)
        #         dfs(r-1,c)
        #         dfs(r, c+1)
        #         dfs(r, c-1)

        # ans = 0
        # for r in range(ROWS):
        #     for c in range(COLS):
        #         if grid[r][c] == "1" and (r,c) not in visited:
        #             ans+=1
        #             dfs(r, c)
        # return ans



        queue = deque()
        visited = set()
        ROWS = len(grid)
        COLS = len(grid[0])
        ans = 0
        directions = [(1,0), (-1,0), (0,-1), (0,1)]

        def isValid(r,c):
            return 0<=r<ROWS and 0<=c<COLS and grid[r][c] == "1"

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1" and (r,c) not in visited:
                    ans +=1
                    queue.append((r,c))
                    visited.add((r,c))
                    while(queue):
                        row, col = queue.popleft()
                        for dr,dc in directions:
                            nr = row + dr
                            nc = col + dc
                            if isValid(nr, nc) and (nr, nc) not in visited:
                                queue.append((nr, nc))
                                visited.add((nr, nc))
        return ans
